### PR TITLE
[lora] support lora for Gemini

### LINKED ISSUE #4929: [gemini] support lora

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

resolved #4929

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

- support lora feature for gemini plugin
- add a method that merges lora weights into original parameters for booster

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_lora/test_gemini_lora.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+            optimizer.step()
+
+    # Check parameters
+    gemini_dict = model.state_dict(only_rank_0=False)
+    model_copy_dict = model_copy.state_dict()
+
+    for name in gemini_dict:
+        check_param_equality(
+            name, gemini_dict[name], model_copy_dict[name], modules_to_save=model.unwrap().modules_to_save
+        )
```
Comment (ver217) [inline tests/test_lora/test_gemini_lora.py:166]: Can you merge these function to single one?

File: colossalai/booster/plugin/gemini_plugin.py [hunk 2]
```diff
@@ -224,6 +224,72 @@ def save_lr_scheduler(self, lr_scheduler: LRScheduler, checkpoint: str):
...
+                )
+            else:
+                torch.save(output_state_dict, os.path.join(checkpoint, WEIGHTS_NAME))
+
+        # save the config and change the inference mode to `True`
+        if peft_config.base_model_name_or_path is None:
+            peft_config.base_model_name_or_path = peft_model.base_model.model.__dict__.get("name_or_path", None)
+
+        inference_mode = peft_config.inference_mode
+        peft_config.inference_mode = True
```
[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/booster/plugin/gemini_plugin.py:273]: Peft will set it to True when saving, and I didn't change this behavior.

File: colossalai/tensor/param_op_hook.py [hunk 3]
```diff
@@ -142,7 +142,6 @@ def _flatten_grad_args(args) -> Tuple[list, list, List[bool], TreeSpec]:
             grad_args.append(arg)
         else:
             other_args.append(arg)
-    assert len(grad_args) > 0
```
[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/tensor/param_op_hook.py:145]: Because there are many parameters not requiring gradients (but they are ColoParameter) when lora is enabled. The operator between parameters without gradients will only produce `other_args` rather than `grad_args`.

[the remark this answers is not in the document]
Comment (Fridge003) [reply colossalai/tensor/param_op_hook.py:145]: They are plenty of parameters with `requires_grad` being False when lora is enabled. So `len(grad_args)` often equals to zero during training.

File: tests/test_lora/test_gemini_lora.py (hunk 1 above)
Comment (Fridge003) [reply tests/test_lora/test_gemini_lora.py]: Done

**Comment by xs1997zju** [comment]
any update of this work?

**Comment by KaiLv69** [comment]
Looking forward to the feature implemented in this PR！

**Comment by ericxsun** [comment]
Also wants this feature. Thanks a lot.
